File: src/main/java/com/extension/factcheck/infrastructure/GoogleIdVerifier.java
package com.extension.factcheck.infrastructure;

import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken;
import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken.Payload;
import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.jackson2.JacksonFactory;
import java.util.Collections;
import java.util.Optional;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
public class GoogleIdVerifier {

  private final NetHttpTransport transport = new NetHttpTransport();

  private final JacksonFactory jsonFactory = JacksonFactory.getDefaultInstance();

  private final GoogleIdTokenVerifier verifier;

  @Value("${chrome.client.id}")
  private String clientId;

  public GoogleIdVerifier() {
    this.verifier = new GoogleIdTokenVerifier.Builder(transport, jsonFactory)
        .setAudience(Collections.singletonList(clientId))
        .build();
  }

  @Transactional
  public Payload verifyAndGetPayload(final String idTokenString) {

    Optional<GoogleIdToken> idTokenOptional;
    try {
      idTokenOptional = Optional.ofNullable(verifier.verify(idTokenString));
    } catch (Exception e) {
      throw new RuntimeException(e);
    }

    GoogleIdToken idToken = idTokenOptional
        .orElseThrow(() -> new RuntimeException("Could not verify Google ID token"));

    return idToken.getPayload();
  }
}

File: src/main/kotlin/com/extension/factcheck/infrastructure/GoogleIdVerifier.kt
package com.extension.factcheck.infrastructure

import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken
import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier
import com.google.api.client.http.javanet.NetHttpTransport
import com.google.api.client.json.gson.GsonFactory
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class GoogleIdVerifier {
    private val transport = NetHttpTransport()
    private val jsonFactory = GsonFactory.getDefaultInstance()

    private val verifier: GoogleIdTokenVerifier = GoogleIdTokenVerifier.Builder(transport, jsonFactory)
        .setAudience(listOf("1056347669242-2agi9u11ic8sjcq299gdch9uadqel7pb.apps.googleusercontent.com"))
        .build()

    @Transactional
    fun verifyAndGetPayload(idTokenString: String): GoogleIdToken.Payload {
        val idToken = try {
            verifier.verify(idTokenString)
        } catch (e: Exception) {
            throw RuntimeException(e)
        } ?: throw RuntimeException("Could not verify Google ID token")

        return idToken.payload
    }
}
